File: pyg4ometry/pygeometry/geant4/solid/Torus.py
from SolidBase import SolidBase as _SolidBase
from pygeometry.geant4.Registry import registry as _registry
from pygeometry.geant4.solid.Wedge import Wedge as _Wedge
from pygeometry.pycsg.core import CSG as _CSG
from pygeometry.pycsg.geom import Vector as _Vector
from pygeometry.pycsg.geom import Vertex as _Vertex
from pygeometry.pycsg.geom import Polygon as _Polygon
import numpy as _np


class Torus(_SolidBase) :
    def __init__(self, name, pRmin, pRmax, pRtor, pSPhi, pDPhi, nslice=16, nstack=16) :
        """
        Constructs a torus. 

        Inputs:
          name:   string, name of the volume
          pRmin:  float, innder radius
          pRmax:  float, outer radius
          pRtor:  float, swept radius of torus
          pSphi:  float, start phi angle
          pDPhi:  float, delta angle
        """
        self.type    = 'Torus'
        self.name    = name
        self.pRmin   = pRmin
        self.pRmax   = pRmax
        self.pRtor   = pRtor
        self.pSPhi   = pSPhi
        self.pDPhi   = pDPhi
        self.nslice = nslice
        self.nstack = nstack
        self.mesh = None
        _registry.addSolid(self)


    def pycsgmesh(self):
#        if self.mesh :
#            return self.mesh

        self.basicmesh()
        self.csgmesh()

        return self.mesh

    def basicmesh(self) :
        polygons = []
        
        dTheta  = 2*_np.pi/self.nstack
        dPhi    = 2*_np.pi/self.nslice
        sphi    = self.pSPhi
        stacks  = self.nstack
        slices  = self.nslice 

        def appendVertex(vertices, theta, phi, r):
            c = _Vector([0,0,0])
            x = r*_np.cos(theta)+self.pRtor
            z = r*_np.sin(theta)
            y = 0
            x_rot = _np.cos(phi)*x - _np.sin(phi)*y
            y_rot = _np.sin(phi)*x + _np.cos(phi)*y
            
            d = _Vector(
                x_rot,
                y_rot,
                z)

            vertices.append(_Vertex(c.plus(d), None))
            
        rinout    = [self.pRmin, self.pRmax]
        self.meshinout = []
        
        for r in rinout:
            for j0 in range(slices):
                j1 = j0 + 0.5
                j2 = j0 + 1
                for i0 in range(stacks):
                    i1 = i0 + 0.5
                    i2 = i0 + 1
                    verticesN = []
                    appendVertex(verticesN, i1 * dTheta, j1 * dPhi + sphi, r)
                    appendVertex(verticesN, i2 * dTheta, j2 * dPhi + sphi, r)
                    appendVertex(verticesN, i0 * dTheta, j2 * dPhi + sphi, r)
                    polygons.append(_Polygon(verticesN))
                    verticesS = []
                    appendVertex(verticesS, i1 * dTheta, j1 * dPhi + sphi, r)
                    appendVertex(verticesS, i0 * dTheta, j0 * dPhi + sphi, r)
                    appendVertex(verticesS, i2 * dTheta, j0 * dPhi + sphi, r)
                    polygons.append(_Polygon(verticesS))
                    verticesW = []
                    appendVertex(verticesW, i1 * dTheta, j1 * dPhi + sphi, r)
                    appendVertex(verticesW, i0 * dTheta, j2 * dPhi + sphi, r)
                    appendVertex(verticesW, i0 * dTheta, j0 * dPhi + sphi, r)
                    polygons.append(_Polygon(verticesW))
                    verticesE = []
                    appendVertex(verticesE, i1 * dTheta, j1 * dPhi + sphi, r)
                    appendVertex(verticesE, i2 * dTheta, j0 * dPhi + sphi, r)
                    appendVertex(verticesE, i2 * dTheta, j2 * dPhi + sphi, r)
                    polygons.append(_Polygon(verticesE))

            mesh      = _CSG.fromPolygons(polygons)
            self.meshinout.append(mesh)
            polygons = []
            
        
        self.mesh = self.meshinout[1]
        return self.mesh


    def csgmesh(self) :
        if self.pRmin != 0:
            self.mesh  = self.meshinout[0].subtract(self.meshinout[1])
        
        else:
           self.mesh = self.meshinout[1].inverse()

        if self.pDPhi != 2*_np.pi:
            wrmax    = 3*self.pRtor #make sure intersection wedge is much larger than solid
            wzlength = 5*self.pRmax
            
            pWedge = _Wedge("wedge_temp",wrmax, self.pSPhi, self.pDPhi, wzlength).pycsgmesh()
            self.mesh = pWedge.intersect(self.mesh)

        return self.mesh

File: pyg4ometry/pygeometry/cpgeant4/solid/Torus.cpp
#include "Torus.h"
#include "Wedge.h"
#include <cmath>

void CSGMesh::TorusAppendVertex(std::vector<Vertex *> &vertices, double theta, double phi, double r,double pRtor) {
  double x = r*cos(theta)+pRtor;
  double z = r*sin(theta);
  double x_rot = cos(phi)*x;
  double y_rot = sin(phi)*x;
  vertices.push_back(new Vertex(new Vector(x_rot,y_rot,z)));
}

CSG* CSGMesh::ConstructTorus(double pRmin,double  pRmax,double  pRtor,double  sphi,double  pDPhi,int slices,int stacks){
  double dTheta = (2.0*M_PI)/double(stacks);
  double dPhi = (2.0*M_PI)/double(slices);

  std::vector<double> rinout(2);
  rinout[0] = pRmin;
  rinout[1] = pRmax;

  std::vector<CSG*> meshinout(2);

  for(int r=0;r<2;r++) {
    std::vector<Polygon*> polygons;
    for(int j0 = 0;j0 < slices;j0++){
      double j1 = j0 + 0.5;
      double j2 = j0 + 1.0;
      for(int i0 = 0;i0 < stacks; i0++){
        double i1 = i0 + 0.5;
        double i2 = i0 + 1.0;
        std::vector<Vertex*> verticesN;
        CSGMesh::TorusAppendVertex(verticesN, i1 * dTheta, j1 * dPhi + sphi, r,pRtor);
        CSGMesh::TorusAppendVertex(verticesN, i2 * dTheta, j2 * dPhi + sphi, r,pRtor);
        CSGMesh::TorusAppendVertex(verticesN, i0 * dTheta, j2 * dPhi + sphi, r,pRtor);
        polygons.push_back(new Polygon(verticesN,NULL));

        std::vector<Vertex*> verticesS;
        CSGMesh::TorusAppendVertex(verticesS, i1 * dTheta, j1 * dPhi + sphi, r,pRtor);
        CSGMesh::TorusAppendVertex(verticesS, i0 * dTheta, j0 * dPhi + sphi, r,pRtor);
        CSGMesh::TorusAppendVertex(verticesS, i2 * dTheta, j0 * dPhi + sphi, r,pRtor);
        polygons.push_back(new Polygon(verticesS,NULL));

        std::vector<Vertex*> verticesW;
        CSGMesh::TorusAppendVertex(verticesW, i1 * dTheta, j1 * dPhi + sphi, r,pRtor);
        CSGMesh::TorusAppendVertex(verticesW, i0 * dTheta, j2 * dPhi + sphi, r,pRtor);
        CSGMesh::TorusAppendVertex(verticesW, i0 * dTheta, j0 * dPhi + sphi, r,pRtor);
        polygons.push_back(new Polygon(verticesW,NULL));

        std::vector<Vertex*> verticesE;
        CSGMesh::TorusAppendVertex(verticesE, i1 * dTheta, j1 * dPhi + sphi, r,pRtor);
        CSGMesh::TorusAppendVertex(verticesE, i2 * dTheta, j0 * dPhi + sphi, r,pRtor);
        CSGMesh::TorusAppendVertex(verticesE, i2 * dTheta, j2 * dPhi + sphi, r,pRtor);
        polygons.push_back(new Polygon(verticesE,NULL));
      }
    }
    meshinout[r] = CSG::fromPolygons(polygons);
  }//end of loop over

  CSG* mesh;
  if(pRmin != 0.){
    mesh = meshinout[0]->Subtract(meshinout[1]);
  }
  else{
    mesh = meshinout[1]->Inverse();
  }
  if(pDPhi != 2.0*M_PI){
    double wrmax = 3.0*pRtor;
    double wzlength = 5.0*pRmax;
    Wedge* wedge_temp = new Wedge("wedge_temp",wrmax,sphi,pDPhi,wzlength);
    CSG* pWedge = wedge_temp->GetMesh();
    mesh = pWedge->Intersect(mesh);
  }
  return mesh;
}
